lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1911,21 +1942,20 @@ static void checkGenericParamAccess(TypeChecker &TC,
   if (minAccessScope.isPublic())
     return;
 
-  // Swift 3.0.0 mistakenly didn't diagnose any issues when the context access
-  // scope represented a private or fileprivate level.
-  if (downgradeToWarning == DowngradeToWarning::No) {
-    if (!accessScope.isPublic() &&
-        !isa<ModuleDecl>(accessScope.getDeclContext()) &&
-        TC.getLangOpts().isSwiftVersion3()) {
-      downgradeToWarning = DowngradeToWarning::Yes;
-    }
+  if (checkUsableFromInline) {
+    auto diag = TC.diagnose(owner,
+                            diag::generic_param_usable_from_inline,
+                            owner->getDescriptiveKind(),
+                            accessControlErrorKind == ACEK::Requirement);
+    highlightOffendingType(TC, diag, complainRepr);
+    return;
   }
 
   auto minAccess = minAccessScope.accessLevelForDiagnostics();
 
   bool isExplicit =
     owner->getAttrs().hasAttribute<AccessControlAttr>() ||
-    owner->getDeclContext()->getAsProtocolOrProtocolExtensionContext();
+    isa<ProtocolDecl>(owner->getDeclContext());
```

Comment:
Oops, good catch! Is there a test for this?

Reply:
Yep, see https://github.com/apple/swift/pull/16586/files#diff-8ace5fe1c031aa385557a1c025b63437